pkg/proc/gdbserial/undo.go

```diff
@@ -2,19 +2,318 @@ package gdbserial
+		return errors.New("Checkpoint note cannot start with a digit.")
+	} else if firstChar == "," || firstChar == "-" || firstChar == "$" {
+		return fmt.Errorf("Checkpoint note cannot start with character: %c", where[0])
+	} else {
+		firstWord := strings.Split(where, " ")[0]
+		reserved := []string{"time", "undo", "redo", "inferior", "start", "annotation",
+			"end", "bookmark", "event", "pc", "wallclock"}
+		for _, reservedWord := range reserved {
+			if firstWord == reservedWord {
+				return fmt.Errorf("Checkpoint note cannot start with reserved word: %s", reservedWord)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Create a Delve checkpoint structure at the current time, with the supplied note.
+func (uc *undoSession) createCheckpoint(p *gdbProcess, where string) (int, error) {
+	err := validateCheckpointNote(where)
+	if err != nil {
+		return -1, err
+	}
+	cpid := uc.checkpointLastId
+	uc.checkpointLastId++
+	when, err := p.conn.undoCmd("get_time")
+	if err != nil {
+		return -1, err
+	}
+	uc.checkpoints[cpid] = proc.Checkpoint{ID: cpid, When: when, Where: where}
+	uc.save(p)
+	return cpid, nil
+}
+
+// Look up a Delve checkpoint structure by time, as represented by the vUDB;get_time serial command.
+func (uc *undoSession) lookupCheckpoint(pos string) (proc.Checkpoint, error) {
+	cpid, _ := strconv.Atoi(pos[1:])
+	checkpoint, exists := uc.checkpoints[cpid]
+	if !exists {
+		return proc.Checkpoint{}, errors.New("Checkpoint not found")
+	}
+	return checkpoint, nil
+}
+
+// Delete a Delve checkpoint structure from our tracking.
+func (uc *undoSession) deleteCheckpoint(p *gdbProcess, id int) {
+	delete(uc.checkpoints, id)
+	uc.save(p)
+}
+
+// Fetch all Delve checkpoint structures and return an array for user display (with the When field
+// rewritten in human readable form).
+func (uc *undoSession) getCheckpoints() ([]proc.Checkpoint, error) {
+	r := make([]proc.Checkpoint, 0, len(uc.checkpoints))
+	for _, cp := range uc.checkpoints {
+		// Convert the internal representation of time (which is based on the serial
+		// protocol level representation) to a human-readable version for display.
+		bbcount, pc, err := undoParseServerTime(cp.When)
+		if err != nil {
+			return nil, err
+		}
+		cp.When = undoTimeString(bbcount, pc)
+		r = append(r, cp)
+	}
+	return r, nil
+}
+
+// Represents a single serialised bookmark in our session file format.
+type bookmarkTime struct {
+	Bbcount uint64 `json:"bbcount"`
+	Pc      uint64 `json:"pc"`
+}
+
+// Represents the overall structure of our session file format.
+type session struct {
+	Bookmarks map[string]bookmarkTime `json:"bookmarks"`
+}
+
+// Get the path to the UDB session file for the current recording.
+func getSessionPath(p *gdbProcess) (string, error) {
+	user, err := user.Current()
+	if err != nil {
+		return "", err
+	}
+	recording_ids, err := p.conn.undoCmd("get_recording_ids")
+	if err != nil {
+		return "", err
+	}
+	uuids := strings.Split(recording_ids, ";")
```

Comment:
I suggest checking that this has the expected number of elements here, as you will be looking up `uuids[1]` later.